runtime/android/core/src/org/xwalk/core/XWalkActivity.java

```diff
@@ -231,22 +232,40 @@ protected void dismissDialog() {
     }
 
     private void getXWalkLibrary() {
-        try {
-            Intent intent = new Intent(Intent.ACTION_VIEW);
-            startActivity(intent.setData(Uri.parse(XWALK_CORE_MARKET_URL)));
-            showDialog(getMarketProgressDialog());
-        } catch (ActivityNotFoundException e) {
-            Log.d(TAG, "Market open failed");
-            showDialog(getMarketOpenFailedDialog());
-            // TODO(sunlin): enable direct download in the future
-            // downloadXWalkLibrary();
+        // The download url is defined by the meta-data item with the name of "xwalk_apk_url"
+        // under the application tag in AndroidManifest.xml. It can also be specified via
+        // --xwalk-apk-url option of make_apk script indirectly.
+        if (mXWalkApkDownloadUrl == null) {
+            try {
+                PackageManager packageManager = getPackageManager();
+                ApplicationInfo appInfo = packageManager.getApplicationInfo(
+                        getPackageName(), PackageManager.GET_META_DATA);
+                mXWalkApkDownloadUrl = appInfo.metaData.getString("xwalk_apk_url");
+                if (mXWalkApkDownloadUrl == null) mXWalkApkDownloadUrl = "";
+                Log.d(TAG, "Download Url: " + mXWalkApkDownloadUrl);
+            } catch (NameNotFoundException | NullPointerException e) {
```

Comment:
Naive question: which of the calls above could throw `NullPointerException`? Isn't that the kind of exception you _want_ to be thrown since it indicates something actually wrong happened?